### PR TITLE
Migrate TransformCorrelatedSingleRowSubqueryToProject to rule and allow it to work with multiple projections

### LINKED ISSUE #11026: Correlated subqueries does not handle coercion
```
presto:tiny> SELECT 1 FROM nation n WHERE 1 = (SELECT n.nationkey);
 _col0
-------
     1
(1 row)

Query 20180712_134539_00036_qa8mt, FINISHED, 1 node
Splits: 20 total, 20 done (100.00%)
0:00 [25 rows, 0B] [243 rows/s, 0B/s]

presto:tiny> SELECT 1 FROM nation n WHERE 1 = (SELECT cast(n.nationkey as decimal(7,2)));
Query 20180712_134545_00037_qa8mt failed: line 1:35: Given correlated subquery is not supported
SELECT 1 FROM nation n WHERE 1 = (SELECT cast(n.nationkey as decimal(7,2)))
```

As you could see above when correlated predicate requires coercion then correlated subquery does not work.

CC: @anusudarsan

### PR DESCRIPTION
**Pull request by atris**
…roject

The commit adds a new rule for $subject, and removes the current optimizer
that performs the task. The new rule works in the same optimizer as InlineProjections
and can deal with duplicate identity projections, allowing subqueries with multiple
projections to work, something the previous optimizer lacked

Fixes #11026 and supersedes #11099

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -316,8 +316,15 @@ public PlanOptimizers(
...
-                new TransformCorrelatedSingleRowSubqueryToProject(),
+                                new ImplementFilteredAggregations(),
+                                new InlineProjections(),
+                                new RemoveRedundantIdentityProjections())),
+                new IterativeOptimizer(
+                        ruleStats,
+                        statsCalculator,
+                        estimatedExchangesCostCalculator,
+                        ImmutableSet.of(
+                                new TransformCorrelatedSingleRowSubqueryToProject())),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: can you run projection related rules within this iterative optimizer?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 2]
```diff
@@ -89,6 +89,7 @@
 import com.facebook.presto.sql.planner.iterative.rule.TransformCorrelatedLateralJoinToJoin;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Commit message is wrong. It should be `Migrate TransformCorrelatedSingleRowSubqueryToProject to IterativeOptimizer Rule`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ *       - single row VALUES()
+ * </pre>
+ * to:
+ * <pre>
+ *   - Project(A, B, C, A + C)
+ *       - (input) plan which produces symbols: [A, B, C]
+ * </pre>
+ */
+
+public class TransformCorrelatedSingleRowSubqueryToProject
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: All migrated rules have to have unit tests. See com.facebook.presto.sql.planner.iterative.rule.TestTransformCorrelatedScalarAggregationToJoin or com.facebook.presto.sql.planner.iterative.rule.TestTransformExistsApplyToScalarLateralJoin

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 4]
```diff
@@ -1,118 +0,0 @@
...
- *     - Project (A + C)
- *       - single row VALUES()
- * </pre>
- * to:
- * <pre>
- *   - Project(A, B, C, A + C)
- *       - (input) plan which produces symbols: [A, B, C]
- * </pre>
- */
-public class TransformCorrelatedSingleRowSubqueryToProject
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java]: Typically during migration we left the legacy rule for couple releases just in case of bug introduction during the migration.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 5]
```diff
@@ -89,6 +89,7 @@
...
 import com.facebook.presto.sql.planner.iterative.rule.TransformCorrelatedScalarAggregationToJoin;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Divide this PR to 2 commits, one with just migration and the other with adding support for query mentioned in the test.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    private static final Pattern<LateralJoinNode> PATTERN = lateralJoin();
+
+    @Override
+    public Pattern<LateralJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LateralJoinNode parent, Captures captures, Context context)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: did you change anything in the code?

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java [hunk 7]
```diff
@@ -197,6 +197,13 @@ public void testCorrelatedScalarSubquery()
                 "VALUES 2");
     }
 
+    @Test
+    public void testCorrelatedSubqueryWithExplicitCoercion()
+    {
+        assertions.assertQuery("SELECT 1 FROM (VALUES 1, 2) t1(b) WHERE 1 = (SELECT cast(b as decimal(7,2)))",
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java]: put each argument in separate line:
```
method(
    arg1,
    arg2);
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 2 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I worded the commit message so since this is not just a migration, but adding more functionality on top. I can reword the message to reflect both the changes here i.e. migration and supporting multiple projection queries.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 3 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Ah, I forgot to add a unit test for cases without multiple projections as well. Will fix, thanks

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 6 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Ugh, unintended relic. Will fix thanks

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: In general try to avoid `adding more`. Try to keep single commit per single change.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 7 above)
Comment (atris) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java]: Ack

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 2 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Ok, will divide the PR

**Comment by kokosing** [comment]
Try to use a review mode when answering to comments or commenting the PR, see https://help.github.com/articles/reviewing-proposed-changes-in-a-pull-request/.

That way you will trigger one email per all your comments. Also you will be able to modify your pending comments during the review.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 1 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Running them in same iterative optimizer runs this rule before InlineProjections, which leads to incorrect results, even if I add InlineProjections before the new rule in the same optimizer.

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Why is so? Rule should not rewrite is is not able to rewrite.

Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I am not sure why. InlineProjections does not fire before TransformCorrelatedSingleRowSubqueryToProjection even if it is added before the new rule in the optimizer

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Iterative optimizer keeps to fire its rules until none of the changed the plan. So `InlineProjections` does not need be fired before `TransformCorrelatedSingleRowSubqueryToProjection`, it could be run after and after that `TransformCorrelatedSingleRowSubqueryToProjection` will get another chance.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: It is slightly hard since the changes are deeply ingrained with each other. Would you mind reviewing this one as is? I will take care of this in further PRs

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java]: No, please revert this change.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: It is not about me, it is more about git history. Having this is in separate commits has many benefits. Imagine that you introduce a bug, like InlineProjections does not work well with subquery rule for some cases we do not know now. In such case we could revert only that change and leave migration as it is.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 4 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java]: Ack, will do

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 5 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Ok, got that. I will rewrite this commit to split accordingly.

**Comment by atris** [comment]
@kokosing Updated per your comments. Please review

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 8]
```diff
@@ -317,7 +317,12 @@ public PlanOptimizers(
                                 new TransformCorrelatedScalarSubquery(), // must be run after TransformCorrelatedScalarAggregationToJoin
                                 new TransformCorrelatedLateralJoinToJoin(),
                                 new ImplementFilteredAggregations())),
-                new TransformCorrelatedSingleRowSubqueryToProject(),
+                new IterativeOptimizer(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: I do not see legacy role injected here

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedSingleRowSubqueryToProject.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                                p.tableScan(ImmutableList.of(p.symbol("nationkey")),
+                                        ImmutableMap.of(p.symbol("nationkey"), new TpchColumnHandle("tpch:nationkey",
+                                                BIGINT))),
+                                p.project(
+                                        Assignments.of(p.symbol("nationkey"), expression("nationkey")),
+                                        p.values(
+                                                ImmutableList.of(),
+                                                ImmutableList.of(
+                                                        ImmutableList.of())))))
+                .matches(project(
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedSingleRowSubqueryToProject.java]: can you assert project expression here

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedSingleRowSubqueryToProject.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        tester().assertThat(new TransformCorrelatedSingleRowSubqueryToProject())
+                .on(p ->
+                        p.lateral(
+                                ImmutableList.of(p.symbol("nationkey"), p.symbol("nationkey")),
+                                p.tableScan(ImmutableList.of(p.symbol("nationkey")),
+                                        ImmutableMap.of(p.symbol("nationkey"), new TpchColumnHandle("tpch:nationkey",
+                                                BIGINT))),
+                                p.project(
+                                        Assignments.of(p.symbol("nationkey"), expression("nationkey")),
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedSingleRowSubqueryToProject.java]: this is identity projection, can you use something more complicated like `a +1`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 11]
```diff
@@ -72,13 +72,28 @@ public Result apply(LateralJoinNode parent, Captures captures, Context context)
                 .where(ProjectNode.class::isInstance)
                 .findAll();
 
-        if (subqueryProjections.size() == 0) {
+        // LateralJoinNode's input's output symbols will be added as identity projections
+        // during building the result ProjectNode. However, InlineProjections might have
+        // added the same identity projection, so only keep one of them
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: but then `RemoveRedundantIdentityProjections` should remove this projection. Please undo these changes.

Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: No, we need this code block because of two reasons:

1) ``RemoveRedundantIdentityProjections`` will remove projections only when both the Projection's assignments and its's source assignments are same. In this case, the Projection's source will have no assignments.

2) The projection being removed in this code block will be added back when this rule constructs its result. So this is technically an idempotent operation. However, this code block is needed to allow ``TransformCorrelatedSingleRowSubqueryToProject`` to fire and actually build the result.

I will re word the comment since we use the term "identity projections" for something slightly different.

**Comment by atris** [comment]
@kokosing Updated

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 8 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: > Do we want to have the legacy rule run in the optimizer?

Yes, we do. Otherwise there is no point to have it in the codebase. We want to run legacy optimizer when iterative optimizer is disabled. Notice that `IterativeOptimizer` accepts a list of legacy optimizers and a set of `rule`s in one of its constructors. When iterative optimizer is disabled then it run legacy rules, otherwise it runs new rules.

I propose you to read carefully how `IterativeOptimizer` is implemented as it is quite important class in Presto optimizer related codebase.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedSingleRowSubqueryToProject.java (hunk 10 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestTransformCorrelatedSingleRowSubqueryToProject.java]: There is no need to use `Ack`. Notice that you sent emails with these messages. We typically use `reactions` for that (see your message were I used one). Then you could use `thumbs up` and I will know that you agree with the comment. Reactions do not send emails.

We receive a lot of emails, I hope that getting one less will make our work a bit more efficient ;).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 12]
```diff
@@ -72,13 +72,28 @@ public Result apply(LateralJoinNode parent, Captures captures, Context context)
                 .where(ProjectNode.class::isInstance)
                 .findAll();
 
-        if (subqueryProjections.size() == 0) {
+        // LateralJoinNode's input's output symbols will be added as in the projections
+        // during building the result ProjectNode. However, InlineProjections might have
+        // added the same projection already, so only keep one of them
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Can you please explain me what projections do you see here. I would love to know why `InlineProjections` and `RemoveRedundantIdentityProjections` do not work here.

Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: @kokosing 

https://github.com/prestodb/presto/pull/11121#discussion_r205733808

**Comment by atris** [comment]
@kokosing Updated per comments

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 8 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: @kokosing Did the same, however, I do not see an example of the constructor being used in ``PlanOptimizers``. Could you please point me to one?

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: It is because we recently we removed them. We were migrating a lot of rules at Teradata, now we do not have so much power to continue on that.

See:
 - a7463e6d96ab417466a29c67a1f33b8b493ab600
 - 9f95e662510606ee383f4c2469dae10aaddb1e45
 - 1816553e6bccaf5bf1a5198ee3269347086fdcb2

**Comment by atris** [comment]
@kokosing For the new test I added, I get the following error in Travis:

com.facebook.presto.spi.PrestoException: Unknown transaction ID: 92c08816-dfed-4ec3-b975-cf2076a48414. Possibly expired? Commands ignored until end of transaction block
	at com.facebook.presto.sql.planner.iterative.rule.TestTransformCorrelatedSingleRowSubqueryToProject.testRewrite(TestTransformCorrelatedSingleRowSubqueryToProject.java:56)

It does not seem something like I have seen before, could you please advise?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 13]
```diff
@@ -72,13 +72,28 @@ public Result apply(LateralJoinNode parent, Captures captures, Context context)
                 .where(ProjectNode.class::isInstance)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Use:
```
where(node -> node instanceof ProjectNode && !node.getOutputSymbols().equals(parent.getCorrelation()))
```

Then please remove all the changes below.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 14]
```diff
@@ -321,7 +321,11 @@ public PlanOptimizers(
                         ruleStats,
                         statsCalculator,
                         estimatedExchangesCostCalculator,
+                        ImmutableList.of(
+                                new com.facebook.presto.sql.planner.optimizations.TransformCorrelatedSingleRowSubqueryToProject()),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: This should go with previous commit, where you migrates the rule. Also please annotate legacy optimizer with `@Deprecated`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 13 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Thanks, this looks much cleaner

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Not only cleaner, it has different semantic due the usage of `parent.getCorrelation()`

Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: Can you help me understand what is the difference between this and the earlier implementation? If I understood correctly, LateralJoin's correlation will always be a subset of its input's output symbols, and in the previous implementation, we were removing the projections whose output symbols are completely contained in parent's input's symbols. I understand that we are ignoring such projections during search itself, rather than removing them afterwards, but is there any other semantic difference, please?

**Comment by kokosing** [comment]
Also you have travis failures which have to be diagnosed.

**Comment by atris** [comment]
@kokosing For the Travis failure, please see: 
https://github.com/prestodb/presto/pull/11121#issuecomment-408813204

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 13 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedSingleRowSubqueryToProject.java]: That is all.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 15]
```diff
@@ -324,6 +324,8 @@ public PlanOptimizers(
                         ImmutableList.of(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: commit message is too long

**Comment by atris** [comment]
@kokosing Fixed

**Comment by kokosing** [comment]
```
[ERROR]   TestNonIterativeDistributedQueries>AbstractTestQueries.testCorrelatedScalarSubqueries:6077->AbstractTestQueryFramework.assertQuery:119->AbstractTestQueryFramework.assertQuery:124 Execution of 'actual' query failed: SELECT max(l.quantity), 2 * l.orderkey FROM lineitem l GROUP BY l.orderkey HAVING max(l.quantity) < (SELECT l.orderkey)
```
Looks like a legit faiulure.

**Comment by atris** [comment]
@kokosing Test failure was an inadvertent change in the PR for a query which was failing earlier (and the check was to ascertain the query continues to fail). I have reverted that change. Please review

**Comment by kokosing** [comment]
Thanks!
